### PR TITLE
recover username rework

### PR DESCRIPTION
**Pull request by pzduniak**
@keybase/y2ksquad 
@keybase/react-hackers

### CODE REVIEW

File: shared/actions/provision.tsx [hunk 1]
```diff
@@ -506,17 +506,31 @@ const showFinalErrorPage = (state: Container.TypedState, action: ProvisionGen.Sh
...
+      {email: action.payload.email},
+      Constants.forgotUsernameWaitingKey
+    )
+      .then(() => ProvisionGen.createForgotUsernameResult({result: 'success'}))
+      .catch(error =>
+        ProvisionGen.createForgotUsernameResult({
+          result: Constants.decodeForgotUsernameError(error),
+        })
+      )
+  } else {
```
Comment (jakob223) [inline shared/actions/provision.tsx]: worth it to check here for a phone and error if there isn't one?

File: shared/provision/forgot-username/index.tsx [hunk 2]
```diff
@@ -1,93 +1,142 @@
...
+  onSubmit: (email: string, phone: string) => void
+  waiting: boolean
 }
 
-type State = {
-  email: string
-}
+const ForgotUsername = (props: Props) => {
+  const [emailSelected, _setEmailSelected] = React.useState(true)
+  const [phoneSelected, _setPhoneSelected] = React.useState(false)
```
Comment (jakob223) [inline shared/provision/forgot-username/index.tsx]: You don't really need two state variables here, do you?

delete `   const [phoneSelected, _setPhoneSelected] = React.useState(false) `, the lines that call `_setPhoneSelected`, and just set `phoneSelected` to `!emailSelected`.

File: shared/provision/forgot-username/index.tsx [hunk 3]
```diff
@@ -1,93 +1,142 @@
...
+      _setEmailSelected(!newState)
+      _setPhoneSelected(newState)
+    },
+    [_setEmailSelected, _setPhoneSelected]
+  )
 
-class ForgotUsername extends React.Component<Props, State> {
-  state = {email: ''}
+  const [email, setEmail] = React.useState('')
+  const [phoneNumber, setPhoneNumber] = React.useState<string | null>(null)
```
Comment (jakob223) [inline shared/provision/forgot-username/index.tsx]: why null and not just empty string? If someone has entered some phone number and then deleted it again, we don't want to be willing to send this.

File: shared/actions/provision.tsx (hunk 1 above)
Comment (pzduniak) [reply shared/actions/provision.tsx]: It's protected by the component's logic, I'll explain this below.

File: shared/provision/forgot-username/index.tsx (hunk 2 above)
Comment (pzduniak) [reply shared/provision/forgot-username/index.tsx]: Good call!

File: shared/provision/forgot-username/index.tsx (hunk 3 above)
Comment (pzduniak) [reply shared/provision/forgot-username/index.tsx]: onChange from the phone number input returns (text, valid) in the callback. If valid is false, setPhoneNumber(null) is executed, therefore the only time when "phoneNumber" is a string, it must be a valid number.

If you think that this is too complicated I can make it an empty string instead.

Comment (jakob223) [reply shared/provision/forgot-username/index.tsx]: Maybe add a comment to that effect somewhere, but sounds fine to me.

**Pull request merged**